substrate/frame/assets/src/lib.rs

```diff
@@ -1819,6 +1856,61 @@ pub mod pallet {
 			)?;
 			Ok(())
 		}
+
+		/// Sets the trusted reserve locations of an asset.
+		///
+		/// Origin must be the Owner of the asset `id`. The origin must conform to the configured
+		/// `CreateOrigin` or be the signed `owner` configured during asset creation.
+		///
+		/// - `id`: The identifier of the asset.
+		/// - `reserves`: The full list of trusted reserves.
+		///
+		/// Emits `AssetMinBalanceChanged` event when successful.
+		#[pallet::call_index(33)]
+		#[pallet::weight(T::WeightInfo::set_reserves())]
+		pub fn set_reserves(
+			origin: OriginFor<T>,
+			id: T::AssetIdParameter,
+			reserves: Vec<T::ReserveId>,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+			// FIXME:
+			// T::ManagerOrigin::ensure_origin(origin, &id)?;
+
+			let id: T::AssetId = id.into();
+			let details = Asset::<T, I>::get(&id).ok_or(Error::<T, I>::Unknown)?;
+			ensure!(origin == details.owner, Error::<T, I>::NoPermission);
+
+			Self::unchecked_update_reserves(id, reserves)?;
+			Ok(())
+		}
+	}
+
+	#[pallet::view_functions]
+	impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		/// Provide the asset details for asset `id`.
+		pub fn asset_details(
+			id: T::AssetId,
+		) -> Option<AssetDetails<T::Balance, T::AccountId, DepositBalanceOf<T, I>>> {
+			Asset::<T, I>::get(id)
+		}
+
+		/// Provide the balance of `who` for asset `id`.
+		pub fn balance_of(who: T::AccountId, id: T::AssetId) -> Option<<T as Config<I>>::Balance> {
+			Account::<T, I>::get(id, who).map(|account| account.balance)
+		}
+
+		/// Provide the configured metadata for asset `id`.
+		pub fn get_metadata(
+			id: T::AssetId,
+		) -> Option<AssetMetadata<DepositBalanceOf<T, I>, BoundedVec<u8, T::StringLimit>>> {
+			Metadata::<T, I>::try_get(id).ok()
+		}
+
+		/// Provide the configured reserves for asset `id`.
+		pub fn get_reserves(id: T::AssetId) -> Vec<T::ReserveId> {
+			Self::reserves(&id)
```

Comment:
I would maybe remove `fn get_reserves` and use `ProvideAssetReserves` when needed, or do it opposite way:
Pallet impl -> DB
ProvideAssetReserves impl -> Pallet impl

```suggestion
		ReserveLocations::<T, I>::get(id).into_inner()
```

Reply:
No because I cannot use reference in view_function so that forces the trait implementation to also use by value, which in turn forces cloning in runtime/xcm filters when checking this, I like to avoid it if possible